### PR TITLE
Utilized: `CheckOwnUID`

### PR DESCRIPTION
**Pull request by dvdvideo1234**
Optimized: Create Vector/Angle for ghosting in every frame

### CODE REVIEW

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 1]
```diff
@@ -100,8 +101,15 @@ local function GetRandomString(nLen)
   return sOut
 end
 
-function TOOL:GetControlUID(sIdx)
-  return SanitizeUID(self:GetClientInfo(sIdx.."uid"))
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:104]: Integrate validation check in the request UID method

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 2]
```diff
@@ -100,8 +101,15 @@ local function GetRandomString(nLen)
...
+function TOOL:GetControlUID(sIdx, bVal)
+  local out = SanitizeUID(self:GetClientInfo(sIdx.."uid"))
+  if ( bVal ) then -- Force validation of UID
+    local ok, err = jcon.isValidUID(out)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:107]: When enabled, forces validation on the UID

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 3]
```diff
@@ -130,76 +138,69 @@ function TOOL:GetNormalSpawn(stTr, eEnt)
   return vNorm, aNorm
 end
 
+function TOOL:CheckOwnUID(sUID, uNtf, bJM)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:141]: This is here to check for ownership instead calling the loop two times in a row on exact same UID

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 4]
```diff
@@ -130,76 +138,69 @@ function TOOL:GetNormalSpawn(stTr, eEnt)
...
+  local ply, stat = self:GetOwner(), 0
+  local wins = jcon and jcon.wireModInstances or nil
+
+  -- Check if the player owns the UID, or if the UID is free
+  if jcon and wins and wins[sUID] then
+    for k, v in pairs(wins[sUID]) do
+      if v == ply then
+        stat = 1
+      elseif ( bJM and sUID == "jm_" ) then
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:150]: Added: Flag for "jm_" and call the case when dedicated code is developed

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 5]
```diff
@@ -130,76 +138,69 @@ function TOOL:GetNormalSpawn(stTr, eEnt)
...
+        stat = 2
+        umsg.Start("joywarn",ply)
+          umsg.Short(uNtf)
+          umsg.String(sUID)
+        umsg.End()
+      end
+    end
+  end
+
+  return stat
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:163]: Call the requested warning `uNtf` and return the status

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 6]
```diff
@@ -130,76 +138,69 @@ function TOOL:GetNormalSpawn(stTr, eEnt)
...
-          status = 2
-          umsg.Start("joywarn",ply)
-            umsg.Short(1)
-            umsg.String(_uid)
-          umsg.End()
-        end
-      end
-    end
+    local stat = self:CheckOwnUID(_uid, 1)
+    if ( stat == 2 and okuid ) then okuid = false end
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:183]: If there is one UID with status 2, return, but warn for all invalids, just line the original author intended

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 7]
```diff
@@ -130,76 +138,69 @@ function TOOL:GetNormalSpawn(stTr, eEnt)
...
-  -- Conflicting UID, exit
-  if status == 2 then return false end
+  -- Some conflicting UID os not OK then exit
+  if ( not okuid ) then return false end
 
   -- Validate and update
   local pass = {}
   for i = 1, 8 do
     local strI = tostring(i)
+    local _uid = self:GetControlUID(strI, true)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:193]: Force the validation instead in the request method

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 8]
```diff
@@ -130,76 +138,69 @@ function TOOL:GetNormalSpawn(stTr, eEnt)
...
-      return false
-    end
+    -- Current UID has been validated
     local _type = self:GetControlType(strI)
     local _description = self:GetControlDescr(strI)
     local _min, _max = self:GetControlBorder(strI)
 
     -- Check if the player owns the UID, or if the UID is free
-    local status = 0
-    if jcon and wins and wins[_uid] then
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:185]: Here again check ownership, but raise different warning

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 9]
```diff
@@ -246,6 +247,7 @@ function TOOL:LeftClick(tr)
 end
 
 function TOOL:RightClick(tr)
+  if CLIENT then return true end
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:250]: This was missing. Client has no stuff to do here :D

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 10]
```diff
@@ -294,12 +296,13 @@ function TOOL:Reload(tr)
   end
 end
 
-function TOOL:UpdateGhostWirejoystick( ent, ply )
+function TOOL:UpdateGhost( ent, ply )
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:299]: Renamed this to be in tone with GM13

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 11]
```diff
@@ -294,12 +296,13 @@ function TOOL:Reload(tr)
...
   if (not ent or not ent:IsValid()) then return end
 
   local tr = ply:GetEyeTrace()
 
   if (not tr.Hit or
+      not tr.Entity or
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:305]: Make sure entity is available befor calling `GetClass` on it

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 12]
```diff
@@ -316,10 +319,10 @@ function TOOL:Think()
   if (not self.GhostEntity or
       not self.GhostEntity:IsValid() or
           self.GhostEntity:GetModel() ~= self.Model) then
-    self:MakeGhostEntity( self.Model, Vector(0,0,0), Angle(0,0,0) )
+    self:MakeGhostEntity( self.Model, gvGhostZero, gaGhostZero )
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:322]: Speed up the ghosting a little but to save FPS

**Review by MattJeanes — APPROVED** [review]
Looks great, thank you!

**Pull request merged**